### DISCUSSION TITLE
Vaapi support testing

### DISCUSSION THREAD

**Discussion by toofar** [qutebrowser/qutebrowser#7917] [Q&A]
Splitting this out from https://github.com/qutebrowser/qutebrowser/issues/5378

With Qt 6.6.0 WebEngine supports [VA-API](https://en.wikipedia.org/wiki/Video_Acceleration_API) for hardware accelerated video decoding. This discussion is to gather resources and test notes to figure out in what situations it works.

**Where can I test?**

There are videos you can test with here: http://crosvideo.appspot.com/

**How can I test?**

* While a video is playing open up `chrome://media-internals/`
* At the top select the relevent video stream (it'll be in state kSuspended or kPlay)
![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/ffcb3aee-9e36-4a61-b23d-82dc896779ef)
* Scroll down and look for `kVideoDecoderName`. If it's using Vaapi I think the decoder should read `VaapiVideoDecoder` (this screenshot is showing a configuration without Vaapi). In some cases you can get the slightly older GPU accelerated decoder `VDAVideoDecoder` to show up too
![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/f51ffc92-b17c-4475-b610-7b35e1a0ab8a)
* Scroll further down to see various logs from the media module.
    * If you have Vaapi support built in but it is not enabled (based on configuration or unsupported codec, and maybe more reasons) you'll see an info message "Cannot select VaapiVideoDecoder for video decoding"
    ![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/c29527a3-188b-41ff-ba10-d1c1b073ca47)
    * If the VaapiDecoder is enabled but there are errors you may see them here too, eg here is what you'll see on AMD graphics:
    ![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/4aabb7d6-770e-4082-b9e3-6d410c920a3c)
    * If you don't see VaapiVideoDecoder mentioned in there at all you probably don't have Vaapi support built in

You can also use `intel_gpu_top` or `nvtop` to monitor the GPU video decoding state, see https://wiki.archlinux.org/title/Hardware_video_acceleration#Verification

**How can I build Vaapi into Qt?**

If you have the development files for libgbm and libva installed the Qt build system should autodetect that and enable Vaapi support. You should see a `VA-API support` entry under `Qt WebEngineCore` in `config.summary`.

**What situations does Vaapi decoding work in?**

* You'll need Qt6.6+
* You probably need Intel graphics, unless you want to patch your chrome sources.
* You may need to pass `enable-features=VaapiVideoDecoder` in qt.args (I don't think that is necessary though)
    * You might also need `ignore-gpu-blocklist`, that's mentioned in the chrome docs, which are a bit out of date
* :warning:  On Linux with desktop OpenGL you need to pass `enable-features=VaapiVideoDecodeLinuxGL` in qt.args. Although I don't think this is a supported configuration
    * TODO: does this even work on intel graphics? On https://bugreports.qt.io/browse/QTBUG-91677 they said you need to pass QT_XCB_GL_INTEGRATION=xcb_egl, assuming egl is OpenGL ES then that implies that desktop OpenGL doesn't work
    * :bulb:  Another option when using desktop OpenGL is to pass `disable-features=UseChromeOSDirectVideoDecoder` in qt.args. This switches away from the Vaapi decoder to the slightly older VDAVideoDecoder which also seems to use hardware acceleration!
* You may be able to set `QT_XCB_GL_INTEGRATION=xcb_egl`
    * This doesn't seem to affect anything for me? It doesn't auto-enable Vaapi and it doesn't make it work when VaapiVideoDecodeLinuxGL is passed
* You can enable the Vulkan backend with `QSG_RHI_BACKEND=vulkan`
    * Note that we haven't tested the vulkan backend well, it may have issues
    * :warning: If you are not on Intel graphics you'll have to pass in `enable-features=VaapiIgnoreDriverChecks`. This isn't a supported configuration (but seems to work with a patch)
* TODO: is any of this relevant for Mac and Windows?
* TODO: does this only work for some codecs? In the source I'm seeing delegates for: av1, h264, h265, vp8, vp9
* TODO: isn't this supposed to be better on Wayland? I'm not sure, the platform feature detection seems to be the same. Beyond that I'm not noticing any difference. Maybe Wayland uses GLES natively but because I don't have that compiled in that's why?

The various platform restrictions are in here: https://github.com/qt/qtwebengine-chromium/blob/a71ba6ffc8e34033b93c0727e8d7eb17f7d6a25f/chromium/media/mojo/services/gpu_mojo_media_client_cros.cc#L96
Although I'm not seeing anything about EGL or GLES in there, so I'm not sure where that comes into it.

There is a bug report here about Vaapi support on Linux in general: https://bugs.chromium.org/p/chromium/issues/detail?id=1236697#c102
Comment 102 seems to have a patch way to get it working on non-intel graphics!

The only slightly out of date Chrome docs on Vaapi are here: https://chromium.googlesource.com/chromium/src/+/refs/heads/main/docs/gpu/vaapi.md

There are various switches which may or may not be useful here: https://github.com/qt/qtwebengine-chromium/blob/a71ba6ffc8e34033b93c0727e8d7eb17f7d6a25f/chromium/ui/gl/gl_switches.cc

Long arch thread about Vaapi and Chrome here: https://bbs.archlinux.org/viewtopic.php?id=244031&p=35

What ubuntu is shipping: https://discourse.ubuntu.com/t/an-overview-of-hardware-acceleration-in-chromium/36672

**More things to test:**

1. Intel graphics:
    1. With a working configuration, which of these qt.args values are needed, if any? `ignore-gpu-blocklist, enable-accelerated-video-decode, enable-features=VaapiVideoDecoder, enable-features=VaapiVideoDecodeLinuxGL, enable-features=VaapiIgnoreDriverChecks`
    2. Does it work with vulkan graphics? (`export QSG_RHI_BACKEND=vulkan`)
    3. Does it work with egl? (`export QT_XCB_GL_INTEGRATION=xcb_egl`)
    4. Does it work under wayland?
    5. Does it work with desktop Open GL? (check `GL_VERSION` in `chrome://gpu`)
2. Non-Intel graphics
    1. Do the patches from the chrome bugtracker work?
        1. The [gist](https://gist.github.com/thubble/235806c4c64b159653de879173d24d9f) works for me to enable Vaapi support on Vulkan on AMD (with the VaapiIgnoreDriverChecks feature passed in), I haven't tried to see if the other three fix desktop GL mode
        2. For Qt 6.7 (currently chrome 118 based) that patch for chrome isn't being updated and the [recommendation](https://gist.github.com/thubble/235806c4c64b159653de879173d24d9f?permalin
k_comment_id=4718214#gistcomment-4718214) is to compile mesa with patches
    3. Even without those patches does adding `disable-features=UseChromeOSDirectVideoDecoder` into qt.args get you hardware acceleration via the `VDAVideoDecoder` decoder when not using vulkan?
4. `xcb_egl`
    1. I don't understand why it's not working for me with this set since the Qt people said it should. Does it only do stuff on intel graphics?
    3. You can tell if this had an effect by looking at `GL_VERSION` in `chrome://gpu`. Without it I'm seeing "4.6 ..." and with it I'm seeing "OpenGL ES 3.2 ..."

**Recommended config so far:**

TODO: turn this into a config.py snippet? Not sure if we can change qt.args in config.py.

```
If not vulkan:
  # This options enables Vaapi on desktop GL, but it doesn't work with 112
  # based (also requires patching chrome for non-intel)
  #enable-features=VaapiVideoDecodeLinuxGL # currently broken
  # So switch back to the older GPU accelerated decoder (which doesn't work
  # with vulkan). Will probably go away in a later chrome version.
  qt.args += disable-features=UseChromeOSDirectVideoDecoder
If vulkan and not intel:
  # Patch chrome and:
  qt.args += enable-features=VaapiIgnoreDriverChecks
```

**Comment by toofar** [comment]
Update on Vulkan support I suppose. I've been re-testing it on my main instances again over the last couple of days and am seeing the same instability on 6.6.0 that I was seeing on 6.5. ~Lots of renderer process crashes (status 11) and~ sometimes if I have two windows open sometimes if I open a background tab in window A with some media on it, and there is media playing in an active tab in window B, both windows will freeze up until I switch tabs away from the one with media playing in window B. I feel like I've been getting a way higher rate of super inconvenient errors in the amdgpu driver too.

Ah, seems the renderer crashes are an [unrelated](https://bugreports.qt.io/browse/QTBUG-82390?focusedId=745718&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-745718) issue.

So back to OpenGL for me, nothing to do with Vaapi I don't think, but I'll be testing VDAVideoDecoder instead.

**Comment by toofar** [comment]
Another update, this one regarding Qt6.7 beta (chrome 118 based).

I can't figure out how to get hardware accelerate video decoding working with an OpenGL backend.
`disable-features=UseChromeOSDirectVideoDecoder` doesn't work anymore, as expected.
`enable-features=VaapiVideoDecodeLinuxGL` successfully selects the VaapiVideoDecode but that logs an error about `failed Initialize()ing the frame pool` (might be an AMD specific issue?)
The [arch megathread](https://bbs.archlinux.org/viewtopic.php?id=244031&p=39) says you need to use angle too, but I get an error when trying that that which seems to say there are no supported angle backends. So we might be waiting on https://bugreports.qt.io/browse/QTBUG-112281 for that. I haven't tried with the patch for that yet. Or maybe I need to add more flags to let it initialize the frame pool again. [This person](https://gist.github.com/thubble/235806c4c64b159653de879173d24d9f?permalink_comment_id=4741858#gistcomment-4741858) was having the same issue as me and they said they got it work on 119. So maybe I'll try again then.

I can get it working easy by enabling the Qt vulkan backend (`export QSG_RHI_BACKEND=vulkan`) ~but in my experience the hardware acceleration from chrome via vulkan (or maybe it's just the Qt vulkan implemenation?) has issues on my AMD 6700 XT: when playing a videos it sometimes uses ~80% of the GPU (not on a fresh instance, not sure what causes that) and some videos have periodic pauses in them (similar to this [report](https://bugs.chromium.org/p/chromium/issues/detail?id=1091969)). The old OpenGL acceleration didn't have this problem.
Edit: the increased GPU load seems to happen after suspending

For AMD users that patch for chrome isn't being updated anymore and [they](https://gist.github.com/thubble/235806c4c64b159653de879173d24d9f?permalink_comment_id=4777977#gistcomment-4777977) recommend rebuilding mesa with a patch or two. [frontends/va: Only export one handle for contiguous planes (!26165) · Merge requests · Mesa / mesa · GitLab](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/26165) and, if on chrome < 119, [Draft: gbm: Allow GBM allocation of YUV surfaces (!23214) · Merge requests · Mesa / mesa · GitLab](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/23214). With them (on 118 based) vulkan is working fine with no patches to chrome/webengine.

I'll keep running with the vulkan backend for now, not 100% convinced in its robustness yet but no crashes in 3 days at least.

**Comment by xaizek** [comment]
Thanks for this guide.  Just wanted to provide a data point on VAAPI working:
 * Slackware current (it's a rolling release, so some older versions of packages also worked)
 * qutebrowser v3.1.0
 * Qt 6.6.2_20240210_15b7e743
 * Mesa 24.0.1
 * xorg-server 21.1.11
 * Intel iGPU of i5-14600K

The only real configuration was setting `QSG_RHI_BACKEND=vulkan` to use Vulkan backend. After using this for about two months, I have only noticed a rare glitch which looks like several pixel rows of rendered text are missing (seems to appear only on some sites) which didn't happen with OpenGL backend.  However, I also have
```python
c.qt.workarounds.disable_accelerated_2d_canvas = 'never'
```
but don't know if it's a workaround for this glitch or something else.  It's a minor thing anyway.
EDIT: the glitch seems to not be related to the workaround option or Vulkan (could be scaling or something else).
EDIT2: no longer works with Qt6 6.7.1, using Vulkan causes a crash.
EDIT3: works fine again with Qt6 6.8.0

**Comment by MingxuanYao0528** [comment]
I'm not an expert in this topic. So I'll just say what I tested so far. Hopefully it could someone.

What I have at the end:
1. In qutebrowser configuration, I have `c.qt.args = ['enable-features=VaapiVideoDecodeLinuxGL']`
2. I have `QT_XCB_GL_INTEGRATION=xcb_egl` in system environment

What I have tested:
1. Even if I have `QT_XCB_GL_INTEGRATION=xcb_egl` in system environment, VaapiVideoDecoder won't be used if I only have `enable-features=VaapiVideoDecoder` AND `ignore-gpu-blocklist` in `c.qt.args`.
2. I also tried  `c.qt.args = ['enable-features=VaapiVideoDecodeLinuxGL']` without setting `QT_XCB_GL_INTEGRATION`. It will end up using  `FFmpegAudioDecoder` after `VaapiVideoDecoder` failed according to the log in `chrome://media-internals/`
![image](https://github.com/user-attachments/assets/ae0f5a14-d572-41ac-8227-c65bad967323)

**Comment by tepo4ka** [comment]
Thanks, got it working!

- Gentoo hardened
- qutebrowser v3.4.0
- QtWebEngine/Qt 6.8.1, based on Chromium 122.0.6261.171
- Mesa 24.3.3
- intel iGPU, i7-8550U
- `qt.args = [ "disable-logging", "enable-gpu-rasterization", "ignore-gpu-blocklist", "enable-features=VaapiVideoDecodeLinuxGL" ]`
- `QT_XCB_GL_INTEGRATION=xcb_egl` was required

![image](https://github.com/user-attachments/assets/8617f6ca-8926-487d-b5ab-e492a933ad6e)
![image](https://github.com/user-attachments/assets/0745efed-97d2-4646-810a-bdf7b789822c)

**Reply by tepo4ka** [reply]
Update 16.02: xcb_egl turned out to be unstable and unusable. First memory usage slowly creeps from ~2-3GB to ~7GB (presumably from watching videos, but usually this happens in the span of 1-2 days), then it crashes without even quotebrowser reporting it.

Now I am trying the Vulkan, and for now it seems good.

- `qt.args = [ "disable-logging", "enable-gpu-rasterization", "ignore-gpu-blocklist" ]`
- `QSG_RHI_BACKEND=vulkan`
- `USE=... vaapi vulkan ...` (Gentoo-specific)

**Reply by Cheesemassacre** [reply]
Looks like `QSG_RHI_BACKEND=vulkan` works on Arch too. Memory usage is the same as without vaapi now. Before I could fill a ram in 30 minutes (16gb).

**Comment by Cheesemassacre** [comment]
Works on my system

- Arch Linux
- qutebrowser V3.4.0
- qt6-webengine 6.8.2-1
- `c.qt.args = [ "disable-logging", "enable-gpu-rasterization", "ignore-gpu-blocklist", "enable-features=VaapiVideoDecodeLinuxGL" ]`
- `QT_XCB_GL_INTEGRATION=xcb_egl ` (doesn't work without it)
- i3 1315u

![1739468026](https://github.com/user-attachments/assets/23b80342-1fb5-4129-8190-ceecb1746f98)

**Reply by tepo4ka** [reply]
Do you have memory leaks or abrupt crashes? I do, and the crash isn't even reported by qutebrowser: first memory usage slowly creeps from ~2-3GB to ~7GB (presumably from watching videos, but usually this happens in the span of 1-2 days), then it crashes. I'm going to try Vulkan this weekend.

**Reply by Cheesemassacre** [reply]
I just switched to qutebrowser after some time and added these options. I need to test it more. So far it's using a lot more ram than brave. No crashes yet.

**Reply by Cheesemassacre** [reply]
Update: I watched a few yt videos and browsed through reddit, twitter... ram usage got to around 10 gb.

**Comment by toofar** [comment]
I'll update the top post once 6.9.1 is out. My limited understanding at this point is:

* X11:
  * Intel: may work out of the box, may need `QSG_RHI_BACKEND=vulkan`, probably doesn't need anything else
  * Nvidia: hmm, not sure. I think it's somewhere between intel and AMD in terms of support
  * AMD: After https://bugreports.qt.io/browse/QTBUG-123607 you should just need `enable-features=VaapiIgnoreDriverChecks enable-features=VaapiVideoDecoder` in qt.args on top of the vulkan backend. On versions without that fix you need to set `QTWEBENGINE_FORCE_USE_GBM=0` too which I think disables hardware acceleration for all non-video content, and is not recommended by upstream
* Wayland:
  * No idea, there's been recent work by upstream for angle on ozone which I think is more related to Wayland than x11. I also think the QT_XCB_GL_INTEGRATION env var is related to wayland? Not sure, not sure what the impact of setting it to xcb is in that case.

Also you can `export QT_LOGGING_RULES=qt.webengine.compositor.debug=true` to get a little more information out of webengine these days.

**Comment by Bzzz-fan** [comment]
I'm using radeon 780m gpu on arch (niri wm, wayland), and both opengl and vulkan backend vaapi works for me

Opengl with qt args `c.qt.args=["ignore-gpu-blocklist","enable-zero-copy","enable-features=CanvasOopRasterization,VaapiIgnoreDriverChecks,VaapiVideoDecodeLinuxGL,FluentOverlayScrollbar"]`, I have no issue with decoding and normal use except it uses too much vram (around 3.5gb out of 4gb).

Vulkan with qt args `c.qt.args=["ignore-gpu-blocklist","enable-zero-copy","enable-features=CanvasOopRasterization,VaapiVideoDecoder,VaapiIgnoreDriverChecks,Vulkan,DefaultANGLEVulkan,VulkanFromANGLE,FluentOverlayScrollbar"]` and qt environment `c.qt.environ={"QTWEBENGINE_FORCE_USE_GBM":"0"}` (the environment variable is necessary, without it the browser will be extremely unstable, and most likely crash with the error `Fatal Python error: Floating-point exception`), on my use case, vulkan backended vaapi introduce some issue on video playback (and this may not be a decoding issue but video filtering issue), it can display a full blue background under certain condition..

UPDATED: I found that by replacing `VaapiVideoDecodeLinuxGL` with `VaapiVideoDecoder`, vulkan is enabled in about:gpu, and then with `c.qt.environ={"QTWEBENGINE_FORCE_USE_GBM":"0"}` vaapi is working pretty well now, without the vram leaking and issues mentioned in the vulkan part above..

**Reply by tepo4ka** [reply]
Seconding `QTWEBENGINE_FORCE_USE_GBM`: I'm on Intel iGPU, and after the recent update (qtwebengine-6.9.1) there were segfaults in web telegram and, for example, [here](https://github.com/qutebrowser/qutebrowser/issues/8622).

**Reply by Bzzz-fan** [reply]
> Seconding `QTWEBENGINE_FORCE_USE_GBM`: I'm on Intel iGPU, and after the recent update (qtwebengine-6.9.1) there were segfaults in web telegram and, for example, [here](https://github.com/qutebrowser/qutebrowser/issues/8622).

It also has issue loading this page on amdgpu with this flag too. However, no pages can be loaded normally on intel iGPU using xe driver without this flag, and the gpu memory leaking issue also persist on my intel platform with i915 driver while `QTWEBENGINE_FORCE_USE_GBM` is commented out.. I'd guess the situation will get better when qtwebengine 6.10 is released for it will based on a newer chromium version and we have different flags for vaapi..

**Comment by Bzzz-fan** [comment]
Updated to qtwebengine 6.10 now with GL vaapi flag `AcceleratedVideoDecodeLinuxGL` ~~seems now having no memory leaking issues, even without using vulkan flags~~ (~~no, it is still right there, and I still need one line like `c.qt.environ={"QSG_RHI_BACKEND":"vulkan","QTWEBENGINE_FORCE_USE_GBM":"0"}
`~~ cannot reproduce now, need further test). Besides, the zero copy flag `AcceleratedVideoDecodeLinuxZeroCopyGL` is not working properly.

By the way, the rendering issue with blurring objects on web with the intel xe driver seems also resolved now (while the xe driver hangs my system while I'm playing videos several times).

~~Cannot find a way to switch to the vulkan backend vaapi now.. it seems QTWEBENGINE_FORCE_USE_GBM=0 is still needed while this option makes vaapi unavailable.~~
It turns out that with additional feature flags `Vulkan,DefaultANGLEVulkan,VulkanFromANGLE` and without the `QTWEBENGINE_FORCE_USE_GBM=0` line, the vulkan backended vaapi then can be still properly enabled.

Still observe a consistent memory leaking when enabled vaapi and after watch some video streams, it seems the memory usage won't go down even I close these tabs.. This is used to be resolved with `QTWEBENGINE_FORCE_USE_GBM=0`, however I cannot find a way to enable vaapi with this env in the recent versions.

Guess it is still some kind of gpu memory leak actually..
<img width="3035" height="1808" alt="Screenshot from 2025-11-07 23-24-32" src="https://github.com/user-attachments/assets/8b5c7236-993e-4db0-8ce7-ef570bafcce3" />

**Comment by Cheesemassacre** [comment]
I use 12th gen integrated intel uhd on X11 and video decoding works fine. No noticeable memory leaks even after few days of not closing the browser. I only use this in my config:

`c.qt.args=["enable-features=AcceleratedVideoDecodeLinuxGL"]`
